langchain/src/experimental/hubs/makersuite/googlemakersuitehub.ts

```diff
@@ -0,0 +1,358 @@
+
+  toModel(): LLM | BaseChatModel {
+    const modelName = this._modelName();
+    const modelSettings = {
+      modelName,
+      ...this.promptData?.runSettings,
+    };
+    if (this.promptType === "chat") {
+      const examples = this._examples();
+      return new ChatGooglePaLM({
+        examples,
+        ...modelSettings,
+      });
+    } else {
+      return new GooglePaLM(modelSettings);
+    }
+  }
+
+  toChain() {
+    return this.toTemplate().pipe(this.toModel() as Runnable);
+  }
+}
+
+interface DriveFileReadParams
+  extends GoogleVertexAIConnectionParams<GoogleAuthOptions> {
+  fileId: string;
+}
+
+interface DriveCallOptions extends AsyncCallerCallOptions {
+  // Empty, I think
+}
+
+interface DriveFileMakerSuiteResponse extends GoogleResponse {
+  data: MakerSuitePromptData;
+}
+
+export class DriveFileReadConnection
+  extends GoogleConnection<DriveCallOptions, DriveFileMakerSuiteResponse>
+  implements DriveFileReadParams
+{
+  endpoint: string;
+
+  apiVersion: string;
+
+  fileId: string;
+
+  constructor(fields: DriveFileReadParams, caller: AsyncCaller) {
+    super(
+      caller,
+      new GoogleAuth({
+        scopes: "https://www.googleapis.com/auth/drive.readonly",
+        ...fields.authOptions,
+      })
+    );
+
+    this.endpoint = fields.endpoint ?? "www.googleapis.com";
+    this.apiVersion = fields.apiVersion ?? "v3";
+
+    this.fileId = fields.fileId;
+  }
+
+  async buildUrl(): Promise<string> {
+    return `https://${this.endpoint}/drive/${this.apiVersion}/files/${this.fileId}?alt=media`;
+  }
+
+  buildMethod(): string {
+    return "GET";
+  }
+
+  async request(
+    options?: DriveCallOptions
+  ): Promise<DriveFileMakerSuiteResponse> {
+    return this._request(undefined, options ?? {});
+  }
+}
+
+export interface CacheEntry {
+  updated: number;
+  prompt: MakerSuitePrompt;
+}
+
+export class MakerSuiteHub {
+  cache: Record<string, CacheEntry> = {};
+
+  cacheTimeout: number;
+
+  caller: AsyncCaller;
+
+  constructor(config?: MakerSuiteHubConfig) {
+    this.cacheTimeout = config?.cacheTimeout ?? 0;
+    this.caller = config?.caller ?? new AsyncCaller({});
+  }
+
+  isValid(entry: CacheEntry): boolean {
+    // If we don't have a record, it can't be valid
+    // And if the cache timeout is 0, we will always refresh, so the cache is invalid
+    if (!entry || this.cacheTimeout === 0) {
+      return false;
+    }
+
+    const now = Date.now();
+    const expiration = entry.updated + this.cacheTimeout;
+    return expiration > now;
+  }
+
+  async forceLoad(id: string): Promise<MakerSuitePrompt> {
+    const fields: DriveFileReadParams = {
+      fileId: id,
+    };
+    const connection = new DriveFileReadConnection(fields, this.caller);
+    const result = await connection.request();
+    const ret = new MakerSuitePrompt(result.data);
+    this.cache[id] = {
+      prompt: ret,
+      updated: Date.now(),
+    };
+    return ret;
+  }
+
+  async load(id: string): Promise<MakerSuitePrompt> {
```

Comment:
Strong preference between `load` vs. `pull`?

Reply:
Changing to pull() and forcePull()